src/main/scala/org/bblfsh/client/libuast/jni_utils.c

```diff
@@ -0,0 +1,241 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "jni_utils.h"
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+// Type signatures; to get the signature of all methods from a class do:
+// javap -s -p SomeClass.class
+// To create a Java version of this module from the same codebase, #ifdefs should
+// be sprinkled here to get the equivalent Java types.
+const char *TYPE_STR = "Ljava/lang/String;";
+const char *TYPE_OPTION = "Lscala/Option;";
+const char *TYPE_SEQ = "Lscala/collection/Seq;";
+const char *TYPE_MAP = "Lscala/collection/immutable/Map;";
+
+// Method signatures. Same as above: use javap to get them from a .class
+const char *METHOD_SEQ_APPLY = "(I)Ljava/lang/Object;";
+const char *METHOD_MAP_APPLY = "(Ljava/lang/Object;)Ljava/lang/Object;";
+const char *METHOD_MUTLIST_TOLIST = "()Lscala/collection/immutable/List;";
+const char *METHOD_MUTLIST_TOIMMLIST = "()Lscala/collection/immutable/List;";
+const char *METHOD_LIST_INIT = "()V";
+const char *METHOD_LIST_PLUSEQ = "(Ljava/lang/Object;)Lscala/collection/mutable/MutableList;";
+const char *METHOD_OPTION_GET = "()Ljava/lang/Object;";
+const char *METHOD_SEQ_SORTED = "(Lscala/math/Ordering;)Ljava/lang/Object;";
+const char *METHOD_MAP_KEYS = "()Lscala/collection/GenIterable;";
+
+// Class fully qualified names
+const char *CLS_NODE = "gopkg/in/bblfsh/sdk/v1/uast/generated/Node";
+const char *CLS_POSITION = "gopkg/in/bblfsh/sdk/v1/uast/generated/Position";
+const char *CLS_ROLE = "gopkg/in/bblfsh/sdk/v1/uast/generated/Role";
+const char *CLS_OPTION = "scala/Option";
+const char *CLS_SEQ = "scala/collection/Seq";
+const char *CLS_MAP = "scala/collection/Map";
+const char *CLS_LIST = "scala/collection/immutable/List";
+const char *CLS_MUTLIST = "scala/collection/mutable/MutableList";
+const char *CLS_ITERABLE = "scala/collection/GenIterable";
+
+extern JavaVM *jvm;
+
+//// JNI helpers
+JNIEnv *getJNIEnv() {
+  JNIEnv *pEnv = NULL;
+
+  switch ((*jvm)->GetEnv(jvm, (void **)&pEnv, JNI_VERSION_1_8))
+  {
+    case JNI_OK:
+      // Thread is ready to use, nothing to do
+      break;
+
+    case JNI_EDETACHED:
+      // Thread is detached, need to attach
+      (*jvm)->AttachCurrentThread(jvm, (void **)&pEnv, NULL);
+      break;
+  }
+
+  return pEnv;
+}
+
+const char *AsNativeStr(jstring jstr) {
+  if (!jstr)
+    return NULL;
+
+  JNIEnv *env = getJNIEnv();
+  if (!env)
+    return NULL;
+
+  const char *tmp = (*env)->GetStringUTFChars(env, jstr, 0);
+  if ((*env)->ExceptionOccurred(env) || !tmp)
+    return NULL;
+
+  // str must be copied to deref the java string before return
+  const char *cstr = strdup(tmp);
+
+  (*env)->ReleaseStringUTFChars(env, jstr, tmp);
+  if ((*env)->ExceptionOccurred(env))
+    return NULL;
+
+  return cstr;
+}
+
+jobject *ToObjectPtr(jobject *object) {
+  jobject *copy = malloc(sizeof(jobject));
```

Comment:
Who takes care of this memory?

Reply:
It's only used in `ChildAt`; in that case the returned node is owned by libuast (which uses it for a call to `CreateXmlNode` (assigns it to the `xmlNodePtr->private` which is freed by libuast's `xmlFreeNode`).

Reply:
`xmlFreeNode` doesn't free `private` pointer (it cannot know whether it should), so I'd say we're leaking memory here.

Reply:
Damn, I just saw the comments about "libxml won't touch it" in the libxml2 doc. Let me investigate how to handle this, since it's a `jobject` returned by the JNI api by value.